zng/docs/zson.md

```diff
@@ -0,0 +1,683 @@
+the `nets` field is an array of networks and illustrates the helpful range of
+primitive types in ZSON.  Note that the syntax here implies
+the type of the array, as it is inferred from the type of the elements.
+
+Finally, there are four more values that show ZSON's efficacy for
+representing metrics.  Here, there are no type decorators as all of the field
+types are implied by their syntax, and hence, the top-level record type is implied.
+For instance, the `ts` field is an RFC 3339 date/time string,
+unambiguously the primitive type `time`.  Further,
+note that the `value` field takes on different types and even a complex record
+type on the last line.  In this case, there is a different type top-level
+record type implied by each of the three variations of type of the `value` field.
+
+## 2. The ZSON Data Model
+
+ZSON data is defined as an ordered sequence of one or more typed data values
+separated at any point or not at all by an end-of-sequence marker `.`.
+Each sequence implies a type context as described below and the `.` marker causes
+a new type context to be established for subsequent elements of the sequence.
+
+Each value's type is either a "primitive type", a "complex type", the "type type",
+or the "null type".
+
+### 2.1 Primitive Types
+
+Primitive types include signed and unsigned integers, IEEE floating point of
+several widths, IEEE decimal,
+string, bstring, byte sequence, boolean, IP address, and IP network.
+
+> Note: The `bstring` type is an unusual mixture of a UTF-8 string
+> with embedded binary data as in
+> [Rust's experimental `bstr` library](https://docs.rs/bstr/0.2.14/bstr/).
+> This type is useful in systems that, for instance, pull data off the network
+> while expecting a string, but sometimes encounter embedded binary data due to
+> bugs, malicious attacks, etc.  It is up to the application to differentiate
+> between a `bstring` value that happens to look like a valid UTF-8 string and
+> an actual UTF-8 string encoded as a `bstring`.
+
+### 2.2 Complex Types
+
+Complex types are composed of primitive types and/or other complex types
+and include
+* _record_ - an ordered collection of zero or more named values called fields,
+* _array_ - an ordered sequence of zero or more values called elements,
+* _set_ - a set of zero or more unique values called elements,
+* _union_ - a type representing values whose type is any of a specified collection of two or more types,
+* _enum_ - a type representing values taken from a specified collection of one or more uniformly
+typed values, which can be either primitive or complex, and
+* _map_ - a collection of zero or more key/value pairs where the keys are of a
+uniform type called the key type and the values are of a uniform type called
+the value type.
+
+### 2.3 The Type Type
+
+ZSON also includes first-class types, where a value can be of type `type`.
+The syntax for type values is given below.
+
+### 2.4 The Null Type
```

Comment:
```suggestion
### 2.4 The Null Type and Null Values
```

Reply:
This breaks the symmetry of the "data model" section describing the various data types.  The format section describes the values.  I will fix this by creating a Null Value section and breaking out the text here into that new section.